### PR TITLE
HPCC-23712 Thor-on-demand with queueing

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

File: ecl/eclagent/eclagent.cpp [hunk 1]
```diff
@@ -3337,6 +3338,12 @@ extern int HTHOR_API eclagent_main(int argc, const char *argv[], StringBuffer *
     else
         agentTopology.setown(createPTree("hthor")); // MORE - this needs thought!
 
+{
```
Comment (richardkchapman) [inline ecl/eclagent/eclagent.cpp]: Is this supposed to be left in?

**Comment by richardkchapman** [comment]
@jakesmith One minor comment

File: ecl/eclagent/eclagent.cpp (hunk 1 above)
Comment (jakesmith) [reply ecl/eclagent/eclagent.cpp]: nope, will removed and squash.

**Comment by jakesmith** [comment]
@richardkchapman - removed tracing code and squashed.

**Comment by richardkchapman** [comment]
@ghalliday Want to take a look?

**Comment by jakesmith** [comment]
@ghalliday - please review

File: dockerfiles/hpcc/values.schema.json [hunk 2]
```diff
@@ -169,6 +169,10 @@
         "replicas": {
           "type": "integer"
         },
+	"maxActive": {
```
Comment (ghalliday) [inline dockerfiles/hpcc/values.schema.json]: minor: indentation doesn't match before/after

File: dockerfiles/hpcc/values.schema.json [hunk 3]
```diff
@@ -206,9 +210,17 @@
           "type": "string",
           "description": "The name of the thor process"
         },
+	"numSlaves": {
```
Comment (ghalliday) [inline dockerfiles/hpcc/values.schema.json]: indent

File: ecl/eclagent/eclgraph.cpp [hunk 4]
```diff
@@ -1506,17 +1510,51 @@ extern IProbeManager *createDebugManager(IDebuggableContext *debugContext, const
 
 void EclAgent::executeThorGraph(const char * graphName)
```
Comment (ghalliday) [inline ecl/eclagent/eclgraph.cpp]: There is similar code in ccdcontext.cpp.  I suspect that needs similar changes.

File: ecl/agentexec/agentexec.cpp [hunk 5]
```diff
@@ -204,29 +206,69 @@ class WaitThread : public CInterfaceOf<IPooledThread>
...
+                runK8sJob(jobSpecName, wuid, jobName, queryComponentConfig().getPropBool("@deleteJobs", true), params);
             }
             else
             {
-                VStringBuffer exec("%s --workunit=%s --daliServers=%s", apptype.str(), wuid.str(), dali.str());
+                VStringBuffer exec("%s --workunit=%s --daliServers=%s", processName.get(), wuid.str(), dali.str());
+                exec.append(" --config=");
+                queryComponentConfig().getProp("@processConfig", exec);
+                if (queryComponentConfig().getPropBool("@useThorQueue", true))
+                    exec.append(" --queue=").append(queue);
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp]: Out of curiosity, what effect does this have?

**Review by ghalliday — COMMENTED** [review]
@jakesmith some small commenst and a question

**Comment by ghalliday** [comment]
Also now needs rebasing..

**Comment by jakesmith** [comment]
@ghalliday - rebased/addressed comments.

**Comment by mckellyln** [comment]
Awesome

**Pull request merged**